### ISSUE TITLE
Terminal widget breaks with Unicode input that cannot be encoded as ASCII

### ISSUE THREAD

**Issue by ntamas** [urwid/urwid#348]
##### Description:

urwid crashes when sending a Unicode character to a Terminal widget with a keypress if the character cannot be encoded as ASCII.

##### Affected versions (if applicable)
- [X] `master` branch (specify commit)
- [X] Latest stable version from `pypi`
- [ ] Other (specify source)

Probably happens only on Python 3.x

##### Steps to reproduce (if applicable)
This is for macOS and Python 3.7:

1. Create a `Terminal` widget
2. Launch a shell in it
3. Switch the keyboard to a layout that includes non-ASCII characters (I was using Hungarian)
4. Press a key that corresponds to a character that is not in the ASCII codetable (I was using "á")

##### Expected/actual outcome
Expected outcome: the keypress should have been sent to the process launched in the Terminal widget.

Actual outcome: a UnicodeEncodeError was thrown from `vterm.py`, line 1627, in: `key = key.encode('ascii')`.

I know that it is not possible to make an educated guess about the encoding to use here, but how about making the encoding a property of the Terminal widget? The encoding could be set to `'ascii'` by default so it won't be a backward-incompatible change. One should also place the `key.encode()` call in a try-catch block or use `key.encode(..., errors='ignore')` to prevent the crash.

Would you accept a PR with these changes?

**Comment by tonycpsu** [comment]
Thanks for reporting.  I've confirmed the issue with the `examples/terminal.py` test program distributed with `urwid.

> Would you accept a PR with these changes?

Absolutely!

**Comment by ntamas** [comment]
Okay, I think I have a working first version now, but I had to realize that in my case it is not enough to use `'utf8'` as the encoding but I also had to call `urwid.set_encoding('utf8')` before startup to make the terminal work; otherwise the characters that I typed were processed correctly by the shell running inside the terminal, but the local echo of the shell got garbled as `urwid` was printing the raw UTF8 bytes back into the terminal one by one.

At this point I was wondering whether we should tie the behaviour of the terminal to the encoding set by the user with `urwid.set_encoding()` instead of relying on the user to specify the encoding once again when the terminal widget is constructed. The reason why I am not so sure about this is because `urwid.set_encoding()` seems to be dealing with _output_ encoding, while the encoding specified here in the terminal widget is the _input_ encoding. Any suggestions? Shall I just submit my PR with the proposed `encoding` argument, or shall we try to make an educated guess based on `urwid.set_encoding()` (which might not be correct).
